File: Review-02-Python-Well-Defined-Interfaces/Example-1/update_room.py
#! /usr/bin/env python

from room import Room


def main():
    """
    Compute the area of a room and the cost of flooring for the room.
    """

    room = Room("Laundry Room", 8, 4, 1.95)
    kitchen = Room("Kitchen", 20, 12, 3.87)

    room_pointer = room

    room.set_flooring("Tile", 2.50)

    print("----------------------------------------------------------------")
    print(room)
    print("----------------------------------------------------------------")
    print(room_pointer)
    print("----------------------------------------------------------------")

    print(
        f"&room ({id(room):#0x}) == &room_pointer ({id(room_pointer):#0x})"
        f" -> {room == room_pointer}"
    )
    print("----------------------------------------------------------------")

    if room == room_pointer:
        print("room == room_pointer")
    else:
        print("room != room_pointer")

    print("----------------------------------------------------------------")

    if room == kitchen:
        print("room == kitchen")
    else:
        print("room != kitchen")

    if room < kitchen:
        print("room < kitchen")
    else:
        print("room >= kitchen")

    print()


if __name__ == "__main__":
    main()

File: Review-02-Java-Well-Defined-Interfaces/Example-1/src/main/java/UpdateRoom.java
import java.io.BufferedReader;
import java.io.StringReader;
import java.io.IOException;
import java.util.Iterator;
import java.util.Arrays;

import house.Room;

public class UpdateRoom
{
    /**
     * Compute the area of a room and the cost of flooring for the room.
     */
    public static void main(String... args)
    {
        Room room = new Room("Laundry Room", 8, 4, 1.95);
        Room kitchen = new Room("Kitchen", 20, 12, 3.87);

        Room roomPointer = room;

        room.setFlooring("Tile", 2.50);

        System.out.println("----------------------------------------------------------------");
        System.out.println(room);
        System.out.println("----------------------------------------------------------------");

        System.out.println(roomPointer);

        System.out.println("----------------------------------------------------------------");
        System.out.printf(
            "&room (%s)  == &roomPointer (%s)   -> %b%n",
            Integer.toHexString(System.identityHashCode(room)),
            Integer.toHexString(System.identityHashCode(roomPointer)),
            (room == roomPointer)
        );
        System.out.println("----------------------------------------------------------------");

        if (room.equals(roomPointer)) {
            System.out.println("room == roomPointer");
        }
        else {
            System.out.println("room != roomPointer");
        }

        System.out.println("----------------------------------------------------------------");

        if (room.equals(kitchen)) {
            System.out.println("room == kitchen");
        }
        else {
            System.out.println("room != kitchen");
        }

        System.out.println();
    }
}
